minimal_high_density.md

```diff
@@ -0,0 +1,103 @@
+Well, we'd like these to be as big as is reasonable,
+because computation will scale with the number of these squares.
+But, this grid is the source of any discretization artifacts:
+the smaller the grid squares are, the closer this model is to `minimal.slim`.
+In both measuring local density and choosing mates we average over a kernel
+with a certain characteristic scale: `SX` for local density (for "interactions") and `SM` for mating.
+So, as long as our grid squares are sufficiently smaller than these,
+any discretization will be smoothed out by those operations.
+So, we set the size of the grid cells to be approximately `min(SX,SM)/2`:
+```
+grid_dims = asInteger(2 * (p1.spatialBounds[c(2,3)] - p1.spatialBounds[c(0,1)]) / min(SX, SM));
+```
+
+The `raw` values are put in two maps: `RAW_DENSITY` and `DENSITY`;
+the first is used for mate choice and the second for interactions.
+
+## Local smoothed density
+
+Local density used as an input for mortality is computed in `minimal.slim`
+by averaging over a Gaussian kernel with standard deviation `SX`.
+To approximate this, we simply smooth the "raw" gridded density by that same kernel:
+```
+DENSITY.smooth(SX * 3, "n", SX);
+```
+and then look up the local, smoothed density at the location of every individual:
+```
+competition = p1.spatialMapValue(DENSITY, inds.spatialPosition);
+```
+This could be made more precise by taking into account that in computing the "raw" density
+also involves some smoothing (the size of the grid squares).
+
+## Mate choice
+
+Mate choice in `minimal.slim` picks a nearby individual proportional to a weight
+assigned from a Gaussian kernel with standard deviation `SM`.
+To approximate this procedure, we choose a nearby location proportional to raw density
+multiplied by the same Gaussian kernel,
+and then return the nearest individual to that location:
+```
+mate_location = RAW_DENSITY.sampleNearbyPoint(individual.spatialPosition, 3 * SM, "n", SM);
+mate = i2.nearestNeighborsOfPoint(mate_location, p1, 1);
+```
+
+## Possible pitfalls
+
+The two lines in the "mate choice" snippet can each cause their own problems if care is not taken.
+
+First, the call to `sampleNearbyPoint` works by rejection sampling:
+it picks a location nearby from the provided kernel,
+and then retains the point with probability proportional to the value of the `RAW_DENSITY` map at that point.
+The way this is implemented means that if values of density in some parts of the map are much higher than in other parts of the map,
+this may do a large number of rejections.
+Concretely, if the density within a circle of radius `3 * SM` around a given individual
+is at most, say $10^{-6}$ times the maximum density over the entire map,
+then choosing a mate for that individual will require sampling millions of locations.
+The solution to this might be biological: first, identify individuals with sufficent possible mates
+to be able to mate successfully; and then only choose mates for those.
```

Comment:
Well, OK, but how would you "identify individuals with sufficient possible mates to be able to mate successfully" if not doing a spatial search of the type we're trying to avoid?  Also, note that "sufficient" is misspelled.

Reply:
Seems like really the guidance ought to be: if density in your simulation varies that dramatically, maybe this approach is not for you.  :->  Also: we could fix this problem, couldn't we?  Assess the max value within the specific grid squares being sampled from and use that max for the rejection sampling?  If that makes sense, maybe you can open an issue on it with whatever discussion you see as needed for me to implement it?